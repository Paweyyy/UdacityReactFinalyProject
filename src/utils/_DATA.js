let users = {
  sarah_edo: {
    id: "sarah_edo",
    name: "Sarah Drasner",
    avatarURL: "https://tylermcginnis.com/would-you-rather/sarah.jpg",
    questions: [],
  },
  tylermcginnis: {
    id: "tylermcginnis",
    name: "Tyler McGinnis",
    avatarURL: "https://tylermcginnis.com/would-you-rather/tyler.jpg",
    questions: [],
  },
  dan_abramov: {
    id: "dan_abramov",
    name: "Dan Abramov",
    avatarURL: "https://tylermcginnis.com/would-you-rather/dan.jpg",
    questions: [],
  }
}

let questions = {
  "xi3ca2jcfvpa0i3t4m7ag": {
    id: "xi3ca2jcfvpa0i3t4m7ag",
    author: "tylermcginnis",
    option_1: "Build your new application with Javascript",
    option_2: "Build your new application with Typescript",
    voted_option_1: ["sarah_edo"],
    voted_option_2: ["tylermcginnis"],
    timestamp: 1518122597860
  },
  "5c9qojr2d1738zlx09afby": {
    id: "5c9qojr2d1738zlx09afby",
    author: "sarah_edo",
    option_1: "Get a office dog",
    option_2: "Get a office cat",
    voted_option_1: ["sarah_edo"],
    voted_option_2: [],
    timestamp: 1518043995650
  }
}

export function _getUsers () {
  return new Promise((res, rej) => {
    setTimeout(() => res({...users}), 1000)
  })
}

export function _getQuestions () {
  return new Promise((res, rej) => {
    setTimeout(() => res({...questions}), 1000)
  })
}

export function _saveQuestionResponse ({ id, optionChosen, authedUser }) {
  return new Promise((res, rej) => {
    setTimeout(() => {
      if(optionChosen === questions[id].option_1){
        questions = {
          ...questions,
          [id]: {
            ...questions[id],
            voted_option_1: [...questions[id].voted_option_1, authedUser]
          }
        }
      }else{
        questions = {
          ...questions,
          [id]: {
            ...questions[id],
            voted_option_2: [...questions[id].voted_option_2,authedUser]
          }
        }
      }
      console.log(questions)
      res()
    }, 500)
  })
}

function generateUID () {
  return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
}

function formatQuestion ({ author, question, option_1, option_2 }) {
  return {
    id: generateUID(),
    author,
    option_1,
    option_2,
    voted_option_1: [],
    voted_option_2: [],
    timestamp: Date.now(),
  }
}

export function _saveQuestion ({ option_1, option_2, author }) {
  return new Promise((res, rej) => {
    const formattedQuestion = formatQuestion({
      option_1,
      option_2,
      author
    })

    setTimeout(() => {
      questions = {
        ...questions,
        [formattedQuestion.id]: formattedQuestion,
      }

      users = {
        ...users,
        [author]: {
          ...users[author],
          questions: users[author].questions.concat([formattedQuestion.id])
        }
      }

      res(formattedQuestion)
    }, 1000)
  })
}
