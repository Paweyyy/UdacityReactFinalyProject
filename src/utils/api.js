import {
  _getUsers,
  _getQuestions,
  _saveQuestionResponse,
  _saveQuestion,
} from './_DATA.js'

export function getInitialData () {
  return Promise.all([
    _getUsers(),
    _getQuestions(),
  ]).then(([users, questions]) => ({
    users,
    questions,
  }))
}

export function saveQuestionResponse (info) {
  return _saveQuestionResponse(info)
}

export function saveQuestion (info) {
  return _saveQuestion(info)
}