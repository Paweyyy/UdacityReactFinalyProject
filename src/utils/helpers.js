export function formatDate (timestamp) {
  const d = new Date(timestamp)
  const time = d.toLocaleTimeString('en-US')
  return time.substr(0, 5) + time.slice(-2) + ' | ' + d.toLocaleDateString()
}

export function formatTweet (tweet, author, authedUser, parentTweet) {
  const { id, likes, replies, text, timestamp } = tweet
  const { name, avatarURL } = author

  return {
    name,
    id,
    timestamp,
    text,
    avatar: avatarURL,
    likes: likes.length,
    replies: replies.length,
    hasLiked: likes.includes(authedUser),
    parent: !parentTweet ? null : {
      author: parentTweet.author,
      id: parentTweet.id,
    }
  }
}

export function formatLeaderboardData({ questions, users }){
    let res = [];
    Object.keys(users).map(user => {
      res = [...res, {
        "user": users[user].id,
        "name": users[user].name,
        "avatarURL": users[user].avatarURL,
        "answered": Object.keys(users[user].answers).length,
        "created": users[user].questions.length
      }]
      return null
    })
    res = res.sort((a,b) => {
      if(a.answered === b.answered){
        return a.created < b.created ? 1 : -1
      }else{
        return a.answered < b.answered ? 1 : -1
      }
    })
    return res
} 