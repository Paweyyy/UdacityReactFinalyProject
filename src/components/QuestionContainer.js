import { connect } from "react-redux"
import * as React from "react";
import { formatDate } from "../utils/helpers";
import { Link } from "react-router-dom";

const QuestionContainer = ({ author, timestamp, id }) => {
    return(
        <div className="question" data-testid="question">
            <div className="questionWrapper">
                <h5>{author}</h5>
                <p className="datetime">{timestamp}</p>
            </div>
            <div className="buttonWrapper">
                <Link className="questionButton" to={`/questions/${id}`}>Show</Link>
            </div>
        </div>
    )
}

const mapStateToProps = ({questions},{id}) => (
    {
        author: questions[id].author,
        timestamp: formatDate(questions[id].timestamp),
        id: id
    }
)

export default connect(mapStateToProps)(QuestionContainer)