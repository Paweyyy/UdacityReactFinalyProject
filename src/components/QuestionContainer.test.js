import { render, fireEvent,find } from "@testing-library/react";
import QuestionContainer from "./QuestionContainer";
import { createStore } from "redux";
import { Provider } from "react-redux";
import reducer from "../reducers";
import { BrowserRouter as Router } from "react-router-dom";
import '@testing-library/jest-dom/extend-expect';
import thunk from "redux-thunk";
import { applyMiddleware } from "redux";
import { getInitialData } from "../utils/api" 

describe("QuestionPage", () => {
    it("will match QuestionContainer", async () => {
        const middleware = applyMiddleware(thunk)

        const initialState = await getInitialData()

        const store = createStore(reducer, initialState, middleware);
    
        var component = render(<Provider store={store}><Router><QuestionContainer id={"8xf0y6ziyjabvozdd253nd"}/></Router></Provider>);
    
        expect(component).toMatchSnapshot();
    })
})