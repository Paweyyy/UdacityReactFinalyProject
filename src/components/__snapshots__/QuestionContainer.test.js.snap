// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionPage will match QuestionContainer 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="question"
        data-testid="question"
      >
        <div
          class="questionWrapper"
        >
          <h5>
            sarahedo
          </h5>
          <p
            class="datetime"
          >
            4:21:AM | 29.6.2016
          </p>
        </div>
        <div
          class="buttonWrapper"
        >
          <a
            class="questionButton"
            href="/questions/8xf0y6ziyjabvozdd253nd"
          >
            Show
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="question"
      data-testid="question"
    >
      <div
        class="questionWrapper"
      >
        <h5>
          sarahedo
        </h5>
        <p
          class="datetime"
        >
          4:21:AM | 29.6.2016
        </p>
      </div>
      <div
        class="buttonWrapper"
      >
        <a
          class="questionButton"
          href="/questions/8xf0y6ziyjabvozdd253nd"
        >
          Show
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
